File: ce-app/frontend/src/api/transcript.ts
import api from './client'

export interface Range { start: number; end: number }

export interface JumpcutResult {
  cuts: Range[]
  keep: Range[]
  duration: number
  removed: number
  kept: number
}

export const transcriptApi = {
  cuts: async (words: Record<string, unknown>[], spans: number[][]) =>
    (await api.post('/transcript/cuts', { words, spans })).data as { cuts: Range[]; removed: number },
  fillers: async (words: Record<string, unknown>[], lang?: string) =>
    (await api.post('/transcript/fillers', { words, lang })).data as { cuts: Range[]; count: number; removed: number },
  jumpcut: async (q: {
    words?: Record<string, unknown>[]
    silences?: Range[]
    remove_fillers?: boolean
    remove_silence?: boolean
    minimum_silence?: number
  }) => (await api.post('/transcript/jumpcut', q)).data as JumpcutResult,
}

File: ce-app/backend/app/routers/transcript.py
"""Tier 1 — transcript-first editing endpoints.

The backend answers "which ranges would this transcript edit remove?"; the editor
applies them as one undoable ripple delete. All inputs are the word timings and
silence ranges the app already measured, so nothing here re-guesses a boundary.
"""
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from core.engine import transcript_edit

router = APIRouter(prefix="/api/transcript", tags=["transcript"])


class CutsRequest(BaseModel):
    words: list[dict] = Field(description="Word timings [{start, end, word/text}]")
    spans: list[list[int]] = Field(description="Inclusive word-index spans to delete")


@router.post("/cuts")
def cuts(payload: CutsRequest) -> dict:
    ranges = transcript_edit.ranges_from_words(payload.words, payload.spans)
    if not ranges and payload.spans:
        raise HTTPException(status_code=404, detail="the selection no longer matches the transcript")
    return {"cuts": ranges, "removed": round(sum(b - a for a, b in ranges), 3)}


class FillersRequest(BaseModel):
    words: list[dict]
    lang: str | None = None


@router.post("/fillers")
def fillers(payload: FillersRequest) -> dict:
    ranges = transcript_edit.filler_ranges(payload.words, payload.lang)
    return {"cuts": ranges, "count": len(ranges),
            "removed": round(sum(b - a for a, b in ranges), 3)}


class JumpcutRequest(BaseModel):
    words: list[dict] = []
    silences: list[dict] = []
    remove_fillers: bool = True
    remove_silence: bool = True
    minimum_silence: float = Field(default=0.4, ge=0.1, le=5.0)


@router.post("/jumpcut")
def jumpcut(payload: JumpcutRequest) -> dict:
    if not payload.words and not payload.silences:
        raise HTTPException(status_code=400, detail="give me words or silence ranges to cut")
    return transcript_edit.jumpcut(
        payload.words, payload.silences,
        remove_fillers=payload.remove_fillers,
        remove_silence=payload.remove_silence,
        minimum_silence=payload.minimum_silence,
    )
